internal/vm/validation.go

```diff
@@ -0,0 +1,225 @@
+package vm
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	v1 "github.com/hashicorp/consul-api-gateway/internal/api/v1"
+	"github.com/hashicorp/consul-api-gateway/internal/common"
+	"github.com/hashicorp/consul-api-gateway/internal/vault"
+	consulapi "github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-multierror"
+)
+
+// Validator is responsible for validating the more complex fields of Gateways and Routes
+type Validator struct {
+	client      *consulapi.Client
+	vaultClient vault.KVClient
+	logger      hclog.Logger
+}
+
+func NewValidator(logger hclog.Logger, vaultClient vault.KVClient, client *consulapi.Client) *Validator {
+	return &Validator{
+		client:      client,
+		vaultClient: vaultClient,
+		logger:      logger,
+	}
+}
+
+// ValidateGateway validates a Gateway definition
+func (v *Validator) ValidateGateway(ctx context.Context, gateway *v1.Gateway) error {
+	var errs multierror.Error
+	listenerNames := map[string]int{}
+	listenerPorts := map[int]int{}
+
+	for i, listener := range gateway.Listeners {
+		// treat unsupplied name as a default name (empty string)
+		name := ""
+		if listener.Name != nil {
+			name = *listener.Name
+		}
+
+		// make sure that each listener has a unique port and name
+		if index, exists := listenerNames[name]; exists {
+			errs.Errors = append(errs.Errors, fmt.Errorf("listener.%d: name %q conflicts with listener.%d", i, name, index))
+		} else {
+			listenerNames[name] = i
+		}
+
+		if index, exists := listenerPorts[listener.Port]; exists {
+			errs.Errors = append(errs.Errors, fmt.Errorf("listener.%d: port %d conflicts with listener.%d", i, listener.Port, index))
+		} else {
+			listenerPorts[listener.Port] = i
+		}
+
+		// validate TLS
+		errs.Errors = append(errs.Errors, v.validateListerTLS(i, listener.Tls)...)
+
+		// validate TLS certs
+		errs.Errors = append(errs.Errors, v.validateListerTLSCertificates(ctx, i, listener.Tls)...)
+	}
+
+	return errs.ErrorOrNil()
+}
+
+func (v *Validator) validateListerTLS(index int, tls *v1.TLSConfiguration) []error {
```

Comment:
Should this be validateListenerTLS? Same question for validateListerTLSCertificates? Logic seems fine though so I'm not gonna block.

Reply:
😊 ... spelling 😞